### PR TITLE
Adds auth0 login functionality to ui & api & better daemon auth

### PR DESCRIPTION
**Pull request by alexandrinaw**
* Ability to log in to frontend app via auth0
  * On the frontend, this is turned on with an env var `VITE_USE_AUTH0`
  * On the server, this is turned on with a feature in `server/Cargo.toml`, `require_auth`
* Adds middleware to check for token in api requests and set the user_id.
* Store user_id in all the appropriate tables
* Change how data is loaded on frontend so we can use hooks from auth0 to get the tokens to include with the requests
* Change how daemon registration & tokens work: doesn't use auth0, but they are added in a more secure way now:
  - Uses access token on initial registration to get client_id and client_secret.
  - The access token is rotatable, old ones become invalid.
  - The access token is tied to the user id, so we know which account the daemon belongs to.
  - daemon stores client_id and client_secret locally, and uses those subsequently.
  - kind of basic security right now -- uses client id & secret in basic auth in header.
  - would be possible to deactivate daemon, but not yet implemented.
  - can extend this further to make things more secure, this seems good enough for now.
 
I still have a list of todos that I keep adding to and subtracting from, but this seems maybe good enough to start testing for now?

### CODE REVIEW

File: pipe/section/section_impls/mycelial_server/src/destination.rs [hunk 1]
```diff
@@ -173,7 +180,10 @@ impl Mycelial {
     }
 
     fn basic_auth(&self) -> String {
```
Comment (wwkeyboard) [inline pipe/section/section_impls/mycelial_server/src/destination.rs:182]: Eventually we should think about pulling all of the auth stuff together, this looks like the same method from http_client.rs.

File: server/migrations/20240205152029_add_user_id_to_tables.sql [hunk 2]
```diff
@@ -0,0 +1,5 @@
+-- Add migration script here
+
+ALTER TABLE pipes ADD COLUMN user_id TEXT DEFAULT '';
```
Comment (wwkeyboard) [inline server/migrations/20240205152029_add_user_id_to_tables.sql:3]: Are any of these used in JOIN's? If so we should keep in the back of our minds that they'll probably need an index, eventually.

**Review by wwkeyboard — APPROVED** [review]
LGTM, my comments are about things well need to do soon but shouldn't block this.

File: myceliald/src/http_client.rs [hunk 3]
```diff
@@ -35,43 +36,79 @@ fn is_for_client(config: &Config, name: &str) -> bool {
...
+        if resp.status() != 200 {
+            return Err(format!(
+                "failed to register client - status code: {:?}",
+                resp.status()
+            )
+            .into());
+        }
+        let provision_client_resp: ProvisionClientResponse = resp.json().await?;
+
+        let mut state = SqliteState::new();
```
Comment (dmzmk) [inline myceliald/src/http_client.rs:98]: http client should have own state management, mixing state of two different things is not a good approach here.

File: myceliald/src/http_client.rs [hunk 4]
```diff
@@ -35,43 +36,79 @@ fn is_for_client(config: &Config, name: &str) -> bool {
...
         }
     }
 
+    // Client should register only once. Subsequently, it should use the client_id and client_secret it gets back from the registration to authenticate itself.
+    async fn register_if_not_registered(&mut self) -> Result<(), SectionError> {
+        let state = self
+            .storage_handle
+            .retrieve_state(std::u64::MAX)
+            .await
+            .unwrap();
```
[the remark this answers is not in the document]
Comment (alexandrinaw) [reply myceliald/src/http_client.rs]: good question. updated where it was easy.

File: server/src/main.rs [hunk 5]
```diff
@@ -341,48 +482,36 @@ async fn log_middleware<B>(
 
 async fn provision_client(
     State(state): State<Arc<App>>,
+    Extension(user_id): Extension<UserID>,
     Json(payload): Json<ProvisionClientRequest>,
 ) -> Result<impl IntoResponse, error::Error> {
     let client_id = payload.client_config.node.unique_id;
+    let unique_client_id = Uuid::new_v4().to_string();
+    let client_secret = Uuid::new_v4().to_string();
```
Comment (dmzmk) [inline server/src/main.rs:490]: is it a good idea to use uuid for client secret?

File: server/migrations/20240205152029_add_user_id_to_tables.sql (hunk 2 above)
Comment (alexandrinaw) [reply server/migrations/20240205152029_add_user_id_to_tables.sql:3]: not used in any joins yet, but they are used in lots of where queries, which can definitely benefit from an index. added!

**Review by dmzmk — CHANGES_REQUESTED** [review]
It's not a good idea for every entity on daemon side to be aware about client_id and client_secret. Only single section requires those values and control plane will not have ingestion endpoints soon enough.

Didn't check server side of things yet.

File: server/migrations/20240205152029_add_user_id_to_tables.sql (hunk 2 above)
Comment (wwkeyboard) [reply server/migrations/20240205152029_add_user_id_to_tables.sql:3]: Thank you!

File: myceliald/src/main.rs [hunk 6]
```diff
@@ -57,7 +58,8 @@ async fn run() -> Result<()> {
 
     let storage_handle = storage::new(config.node.storage_path.clone()).await?;
     let runtime_handle = runtime::new(storage_handle.clone());
-    let client_handle = http_client::new(config, runtime_handle, storage_handle);
+    let daemon_storage = daemon_storage::new("./.mycelial_daemon".to_string()).await?; // no idea if this is a good spot to store this info...
```
Comment (dmzmk) [inline myceliald/src/main.rs:61]: good enough for now. Feels like in near future we can use daemon's database to store this info

**Review by dmzmk — APPROVED** [review]
lgtm

**Pull request merged**